### ISSUE TITLE
CFE doesn't check type arguments against the bounds of the corresponding type variables

### ISSUE THREAD

**Issue by stefantsov** [dart-lang/sdk#33308]
Consider the following script:

```dart
void foo<T extends num>(T t) {}

class Foo<U extends num> {}

main() {
  foo("bar");
  new Foo<String>();
}
```

Both lines in `main` should result in a compile-time error, because `String` is not a subtype of `num`.  However, currently it is compiled into the following without any error messages:

```kernel
library;
import self as self;
import "dart:core" as core;

class Foo<U extends core::num = core::num> extends core::Object {
  synthetic constructor •() → void
    : super core::Object::•()
    ;
}
static method foo<T extends core::num = core::num>(self::foo::T t) → void {}
static method main() → dynamic {
  self::foo<core::String>("bar");
  new self::Foo::•<core::String>();
}
```

**Comment by stefantsov** [comment]
I've started working on it in [CL 57621](https://dart-review.googlesource.com/c/sdk/+/57621).

**Comment by stefantsov** [comment]
Blocked on https://github.com/dart-lang/sdk/issues/32130.

**Comment by dgrove** [comment]
with -dev.59 analyzer:

cat ~/tmp/dart2stable/33308.dart 
void foo<T extends num>(T t) {}

class Foo<U extends num> {}

main() {
  foo("bar");
  new Foo<String>();
}

dgrove@dgrove:~$ /usr/lib/google-dartlang/bin/dartanalyzer --preview-dart-2 ~/tmp/dart2stable/33308.dart 
Analyzing /usr/local/google/home/dgrove/tmp/dart2stable/33308.dart...
  error • Couldn't infer type parameter 'T'.

Tried to infer 'String' for 'T' which doesn't work:
  Type parameter 'T' declared to extend 'num'.
The type 'String' was inferred from:
  Parameter 't' declared as     'T'
                but argument is 'String'.

Consider passing explicit type argument(s) to the generic.

 at tmp/dart2stable/33308.dart:6:3 • strong_mode_could_not_infer
  error • 'String' doesn't extend 'num' at tmp/dart2stable/33308.dart:7:11 • type_argument_not_matching_bounds
2 errors found.

**Comment by eernstg** [comment]
I closed issue #33585 as a duplicate of this one, but note that it contains further examples that may be useful here.

**Comment by eernstg** [comment]
Similarly closed #33624 as a duplicate of this issue; note that #33624 contains additional examples as well.

**Comment by eernstg** [comment]
Issue #33626 reports on a missing bounds violation error in the analyzer, but also mentions that the same error is missing when the program is run by the VM. Hence, the example in that issue should be taken into account when dealing with this issue.

**Comment by askeksa-google** [comment]
Closed https://github.com/dart-lang/sdk/issues/33928 as a duplicate. Example involves checking type arguments against each other when a bound refers to a sibling type variable.

**Comment by askeksa-google** [comment]
Closed https://github.com/dart-lang/sdk/issues/33863 as a duplicate. This involves a type argument inferred from a function argument which does not satisfy the bound.

**Comment by askeksa-google** [comment]
Closed https://github.com/dart-lang/sdk/issues/33071 as a duplicate. Invalid bounds in mixins.

**Comment by askeksa-google** [comment]
Closed https://github.com/dart-lang/sdk/issues/33018 as a duplicate. Type arguments to class instantiations.

**Comment by askeksa-google** [comment]
Closed https://github.com/dart-lang/sdk/issues/31472 as a duplicate. Type arguments to class instantiations.

**Comment by askeksa-google** [comment]
Closed https://github.com/dart-lang/sdk/issues/32723 as a duplicate. Involves a generic tear-off bounded by a class type parameter.

**Comment by stefantsov** [comment]
More test cases for the issue could be found in https://github.com/dart-lang/sdk/issues/34230.

**Comment by stefantsov** [comment]
Prerequisites for the fix were recently landed, and I'm currently continue the work in [CL 57621](https://dart-review.googlesource.com/c/sdk/+/57621).

**Comment by leafpetersen** [comment]
Will this example be covered by this bug or should I file a separate issue:

```dart
class Foo<T> {}

class Bar<T extends Foo<T>> {}

class Baz extends Bar {}

void main() {
}
```

**Comment by stefantsov** [comment]
@leafpetersen Yes, it's going to be covered.

**Comment by kmillikin** [comment]
Elevating the priority to ensure we ship this with Dart 2.1.
